File: nba_standing.ts
const input =
	'[{"home_team":"Cav","away_team":"Wash","winning_team":"Cav"},{"home_team":"Wash","away_team":"Cav","winning_team":"Wash"},{"home_team":"Golden","away_team":"Wash","winning_team":"Golden"},{"home_team":"Cav","away_team":"Golden","winning_team":"Cav"},{"home_team":"Cav","away_team":"Wash","winning_team":"Cav"}]';

function NBAStanding(input: string) {
	const inp: { home_team: string; away_team: string; winning_team: string }[] =
		JSON.parse(input);
	const map = new Map<string, { wins: number; losses: number }>();

	for (const e of inp) {
		const { home_team, away_team, winning_team } = e;

		if (!map.has(home_team)) {
			map.set(home_team, { wins: 0, losses: 0 });
		}

		if (!map.has(away_team)) {
			map.set(away_team, { wins: 0, losses: 0 });
		}

		const ht_record = map.get(home_team);
		const at_record = map.get(away_team);

		if (!ht_record || !at_record) throw Error('Illegal State');

		if (winning_team == home_team) {
			ht_record.wins++;
			at_record.losses++;
		} else {
			at_record.wins++;
			ht_record.losses++;
		}
	}

	// loop through each key and val
	const res: { team: string; wins: number; losses: number }[] = [];

	for (const team of map.keys()) {
		const record = map.get(team);
		if (!record) throw Error('Invalid State');
		res.push({ team, wins: record.wins, losses: record.losses });
	}

	return res;
}

NBAStanding(input);

File: nba_standing.py
import json
import pprint


class Solution:
    # return { wins: int, loses: int, team }
    def parseStanding(self, input):
        matches = json.loads(input)
        pprint.pprint(matches)
        res = {}

        for m in matches:
            home_team, away_team, winning_team = (
                m["home_team"],
                m["away_team"],
                m["winning_team"],
            )
            if not home_team in res:
                res[home_team] = {"wins": 0, "loses": 0}
            if not away_team in res:
                res[away_team] = {"wins": 0, "loses": 0}

            res[winning_team]["wins"] += 1
            if winning_team == away_team:
                res[home_team]["loses"] += 1
            else:
                res[away_team]["loses"] += 1

        print(res)
        return [
            {"team": k, "wins": res[k]["wins"], "loses": res[k]["loses"]} for k in res
        ]


input = '[{"home_team":"Cav","away_team":"Wash","winning_team":"Cav"},{"home_team":"Wash","away_team":"Cav","winning_team":"Wash"},{"home_team":"Golden","away_team":"Wash","winning_team":"Golden"},{"home_team":"Cav","away_team":"Golden","winning_team":"Cav"},{"home_team":"Cav","away_team":"Wash","winning_team":"Cav"}]'
Solution().parseStanding(input)
